### PR TITLE
Add ZLG-CAN and Tosun CAN device

### PR DESCRIPTION
**Pull request by zhuyu4839**
Add ZLG USB CAN device such as USBCANFD-200U and Tosun USB CAN device such as TC1001, TC1011, TC1004, TC1014, TC1016 etc. support. Including TSMaster COM driver

### CODE REVIEW

**Comment by zhuyu4839** [comment]
see  zhuyu4839/zlgcan-driver: ZLGCAN driver (github.com)
<https://github.com/zhuyu4839/zlgcan-driver>

Alian ***@***.***> 于2022年10月11日周二 16:50写道：

**Comment by zhuyu4839** [comment]
> from zlgcan import ZCAN, ZCANDeviceType, ZCAN_Transmit_Data, ZCAN_TransmitFD_Data, ZCAN_Receive_Data, ZCAN_ReceiveFD_Data, ZCANDataObj, ZCANException, ZCANMessageType, ZCANCanTransType
> 
> My zlgcan file doesnot contain "ZCANDeviceType, ZCAN_Transmit_Data, ZCAN_TransmitFD_Data, ZCAN_Receive_Data, ZCAN_ReceiveFD_Data, ZCANDataObj, ZCANException, ZCANMessageType, ZCANCanTransType". Have you changed the zlgcan.py file?

Please see https://github.com/zhuyu4839/zlgcan-driver

File: can/interfaces/tosun.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+import warnings
+from typing import List, Optional, Tuple, Union, Deque, Any
+
+import can
+import can.typechecking
+import ctypes
+from can import CanInitializationError
+from can.bus import LOG
+
+from tosun import TSCanMessage, TSCanFdMessage, TSMasterException, TosunDevice, TSMasterMessageType
```
Comment (hardbyte) [inline can/interfaces/tosun.py:12]: Suggest adding an entry for the optional dependency `tosun` in `extras_require` in `setup.py`.

**Review by hardbyte — CHANGES_REQUESTED** [review]
Thanks for the contribution to python-can!

Please add just one interface in a PR.

Please also add docs where appropriate. At a minimum add to `doc/interfaces.rst` and add
  a new interface specific document in `doc/interface/tosun` and/or `doc/interface/zlg`.

Testing is always a good idea, hard for software that requires hardware but have a look at how other interfaces have managed it.

Suggest making this PR focus on tosun and working with @keelung-yang to take the best from this PR's zlgcan implementation and https://github.com/hardbyte/python-can/pull/1209.

**Comment by zariiii9003** [comment]
I see that your library already implements the BusABC API. You could use the [plugin interface](https://python-can.readthedocs.io/en/stable/plugin-interface.html) so your `ZCanBus` can be used via python-can.

File: can/interfaces/tosun.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+import ctypes
+from can.bus import LOG
+from can.exceptions import (
+    CanError,
+    CanInterfaceNotImplementedError,
+    CanOperationError,
+    CanInitializationError,
+)
+
+from tosun import TSCanMessage, TSCanFdMessage, TSMasterException, TosunDevice, TSMasterMessageType
```
Comment (notmmaoo) [inline can/interfaces/tosun.py:17]: > from tosun import TSCanMessage, TSCanFdMessage, TSMasterException, TosunDevice, TSMasterMessageType 
How can I find/install `tosun` package?

**Comment by notmmaoo** [comment]
> from tosun import TSCanMessage, TSCanFdMessage, TSMasterException, TosunDevice, TSMasterMessageType
@zhuyu4839 
How do I find and install the `tosun` package?

**Comment by luojiaaoo** [comment]
I really hope you can pr successfully, because tosun and ZLG are very popular in China😄

> > from tosun import TSCanMessage, TSCanFdMessage, TSMasterException, TosunDevice, TSMasterMessageType
> > @zhuyu4839
> > How do I find and install the `tosun` package?

The package is not part of pypi, need install TSMaster software,  and use it to install tosun,

File: can/interfaces/tosun.py (hunk 2 above)
Comment (zhuyu4839) [reply can/interfaces/tosun.py:17]: The tosun package is planing to be refactored by using rust

File: can/interfaces/tosun.py (hunk 1 above)
Comment (zhuyu4839) [reply can/interfaces/tosun.py:17]: The ZLG-CAN device has been refactored by using rust and upload to pypi.
[https://pypi.org/project/zlgcan-driver-py/](url)

**Pull request closed without merging**